### PR TITLE
feat: add synthetic HTS transaction support to debug_traceTransaction & debug_traceBlockByNumber

### LINKED ISSUE #4672: Add Synthetic Traces to debug_traceBlockByNumber
## User Story

**As a** blockchain indexer or developer using the JSON-RPC relay

**I want** `debug_traceBlockByNumber` to return traces for all transactions in a block, including synthetic HTS token transfers

**So that** I can reliably index all transactions without getting stuck when the trace count doesn't match the transaction count from `eth_getBlockByNumber`

## Context

Currently, `debug_traceBlockByNumber` only returns traces for real EVM transactions (those with contract results in Mirror Node). However, `eth_getBlockByNumber` includes synthetic transactions representing HTS token transfers via the `populateSyntheticTransactions` method.

This inconsistency causes:

- Indexers to get stuck when they expect traces for all transactions in a block
- Different RPC methods returning different counts for the same block
- Breaking the Ethereum JSON-RPC invariant that every transaction in a block should be traceable

## Technical Requirements

- [ ] Fetch logs in addition to contract results in `traceBlockByNumber` (similar to `eth_getBlockByNumber`)
- [ ] Create a `populateSyntheticTraces` method in `DebugImpl` that mirrors the pattern of `BlockService.populateSyntheticTransactions`
- [ ] For CallTracer: Return minimal trace structure for synthetic transactions with:
  - `type`: 'CALL'
  - `from`/`to`: Extracted from log topics (Transfer event) or fallback to `log.address`
  - `gas`: `numberTo0x(constants.TX_DEFAULT_GAS_DEFAULT)`
  - `gasUsed`: `constants.ZERO_HEX`
  - `value`: `constants.ZERO_HEX`
  - `input`/`output`: `constants.EMPTY_HEX`
  - `calls`: `[]`
- [ ] For PrestateTracer: Return empty state map `{}` for synthetic transactions
- [ ] Update the early return condition to check for both empty contract results AND empty logs
- [ ] Add unit tests for `populateSyntheticTraces` method
- [ ] Add integration tests verifying consistency between `eth_getBlockByNumber` and `debug_traceBlockByNumber`
- [ ] Add test for blocks containing only HTS operations (should return synthetic traces, not empty array)

### Scenario 1: Block with only synthetic transactions

**Given** a block that contains only HTS token transfers (no EVM contract calls)

**When** I call `debug_traceBlockByNumber` with the CallTracer

**Then** I receive an array of trace results matching the transaction count from `eth_getBlockByNumber`, with each synthetic transaction having a minimal trace structure

---

### Scenario 2: Block with mixed transactions

**Given** a block that contains both real EVM transactions and synthetic HTS transfers

**When** I call `debug_traceBlockByNumber` with the CallTracer

**Then** I receive traces for all transactions:

- Real EVM transactions have full trace data from Mirror Node actions endpoint
- Synthetic transactions have minimal trace structures

---

### Scenario 3: Consistency with eth_getBlockByNumber

**Given** any block on mainnet or testnet

**When** I call both `eth_getBlockByNumber` and `debug_traceBlockByNumber`

**Then** the number of transactions returned by `eth_getBlockByNumber` equals the number of traces returned by `debug_traceBlockByNumber`

### LINKED ISSUE #4673: Add Synthetic Transaction Support to debug_traceTransaction
## User Story

**As a** blockchain developer or indexer using the JSON-RPC relay

**I want** `debug_traceTransaction` to return a trace for synthetic HTS token transfer transactions

**So that** I can trace any transaction hash that appears in `eth_getBlockByNumber`, without receiving errors for synthetic transactions

## Context

Currently, `debug_traceTransaction` throws a `RESOURCE_NOT_FOUND` error when called with a synthetic transaction hash (HTS token transfer). This is because synthetic transactions don't have entries in Mirror Node's `/contracts/results/{hash}` or `/contracts/results/{hash}/actions` endpoints.

However, since synthetic transactions appear in `eth_getBlockByNumber` responses (to maintain HTS/ERC-20 compatibility), users expect to be able to trace them. When `debug_traceBlockByNumber` doesn't return a trace for a transaction, indexers often fall back to `debug_traceTransaction` - which also fails for synthetic transactions, completely blocking the indexer.

## Technical Requirements

- [ ] Update `callTracer` method to handle missing contract results gracefully
- [ ] When no contract result is found, query logs by transaction hash to check if it's a synthetic transaction
- [ ] If logs exist for the transaction hash, return a minimal trace structure instead of throwing an error
- [ ] Minimal trace structure should match `debug_traceBlockByNumber` synthetic traces:
  - `type`: 'CALL'
  - `from`/`to`: Extracted from log topics (Transfer event) or fallback to `log.address`
  - `gas`: `numberTo0x(constants.TX_DEFAULT_GAS_DEFAULT)`
  - `gasUsed`: `constants.ZERO_HEX`
  - `value`: `constants.ZERO_HEX`
  - `input`/`output`: `constants.EMPTY_HEX`
  - `calls`: `[]`
- [ ] Only throw `RESOURCE_NOT_FOUND` if both contract results AND logs are not found
- [ ] Consider updating `prestateTracer` similarly if needed
- [ ] Add unit tests for synthetic transaction handling in `callTracer`
- [ ] Add integration tests for tracing synthetic transaction hashes

### Scenario 1: Trace a synthetic HTS transfer

**Given** a synthetic transaction hash from an HTS token transfer that appears in `eth_getBlockByNumber`

**When** I call `debug_traceTransaction` with the CallTracer on that hash

**Then** I receive a minimal trace structure (not an error), containing the transaction type, addresses, and empty call data

---

### Scenario 2: Non-existent transaction still returns error

**Given** a transaction hash that doesn't exist (not in contract results AND not in logs)

**When** I call `debug_traceTransaction` with that hash

**Then** I receive a `RESOURCE_NOT_FOUND` error as expected

---

### Scenario 3: Real EVM transaction unchanged

**Given** a real EVM transaction hash (with contract results in Mirror Node)

**When** I call `debug_traceTransaction` with the CallTracer

**Then** I receive the full trace data from Mirror Node actions endpoint (existing behavior unchanged)

---

### Scenario 4: Indexer fallback works

**Given** an indexer that falls back to `debug_traceTransaction` when `debug_traceBlockByNumber` returns fewer traces than expected

**When** the indexer calls `debug_traceTransaction` on a synthetic transaction hash

**Then** the indexer receives a valid trace and can proceed without getting stuck

### PR DESCRIPTION
**Pull request by quiet-node**
### Description

This PR adds comprehensive support for synthetic HTS token transfer transactions in both `debug_traceTransaction` and `debug_traceBlockByNumber`. Synthetic transactions (HTS transfers) appear in `eth_getBlockByNumber` but lack contract results in Mirror Node. This PR enables tracing them by:

1. **debug_traceTransaction**: Falls back to log-based minimal traces when contract results are unavailable
2. **debug_traceBlockByNumber**: Fetches both contract results (EVM) and logs (synthetic) to capture all transactions in a block

### Related issue(s)

Fixes #4672
Fixes #4673

### Testing Guide

1. Create HTS token and perform native transfer to generate synthetic transaction.
2. Call `debug_traceTransaction` with synthetic hash, or `debug_traceBlockByNumber` with blockNumber using
    - CallTracer; verify minimal trace returned (no error).
    - Test PrestateTracer (empty object) and OpcodeLogger (minimal result).
3. Confirm non-existent hashes still throw `RESOURCE_NOT_FOUND`.
4. Run unit and acceptance tests for synthetic handling.

### Changes from original design (optional)

Issue #4673 focused on CallTracer with optional PrestateTracer. This PR's implementation extends to all tracers (CallTracer, PrestateTracer, OpcodeLogger) for consistency.

### Additional work needed (optional)

N/A

### Checklist

- [x] I've assigned an assignee to this PR and related issue(s) (if applicable)
- [x] I've assigned a label to this PR and related issue(s) (if applicable)
- [x] I've assigned a milestone to this PR and related issue(s) (if applicable)
- [x] I've updated documentation (code comments, README, etc. if applicable)
- [x] I've done sufficient testing (unit, integration, etc.)

### CODE REVIEW

File: packages/relay/src/lib/debug.ts [hunk 1]
```diff
@@ -635,4 +645,66 @@ export class DebugImpl implements Debug {
...
+    requestDetails: RequestDetails,
+  ): Promise<any> {
+    const logs = await this.common.getLogsWithParams(null, { 'transaction.hash': transactionIdOrHash }, requestDetails);
+
+    if (logs.length === 0) {
+      throw predefined.RESOURCE_NOT_FOUND(`Failed to retrieve transaction information for ${transactionIdOrHash}`);
+    }
+
+    const log = logs[0];
+    switch (tracer) {
```
Comment (simzzz) [inline packages/relay/src/lib/debug.ts]: nit: should we also add a default case for completeness?

**Review by natanasow — APPROVED** [review]

**Review by simzzz — APPROVED** [review]
